File: queue.go
package main

import "fmt"

type QNode struct {
    value string
    next  *QNode
}

type Queue struct {
    head *QNode
    tail *QNode
}

func queueCreateNode(data string) *QNode {
    return &QNode{value: data, next: nil}
}

func createQueue() *Queue {
    return &Queue{head: nil, tail: nil}
}

func enqueue(queue *Queue, data string) {
    newNode := queueCreateNode(data)
    if queue.head == nil {
        queue.head = newNode
        queue.tail = newNode
        return
    }

    queue.tail.next = newNode
    queue.tail = newNode
}

func dequeue(queue *Queue) {
    if queue.head == nil {
        fmt.Println("Очередь пустая. Нечего удалять.")
        return
    }

    if queue.head == queue.tail {
        queue.head = nil
        queue.tail = nil
    } else {
        queue.head = queue.head.next
    }
}

func printQueue(queue *Queue) {
    if queue.head == nil {
        fmt.Println("Очередь пуста.")
        return
    }

    fmt.Print("Очередь: ")
    curElem := queue.head
    for curElem != nil {
        fmt.Print(curElem.value + " ")
        curElem = curElem.next
    }

    fmt.Println()
}

File: queue.cpp
#include <iostream>
#include <string>
#include <fstream>
#include "queue.h"

using namespace std;

QNode* createNode(string data) {
    QNode* newNode = new QNode{data,nullptr};
    return newNode;
}

void createQueue(Queue* queue) {
    queue->head = nullptr;
    queue->tail = nullptr;
}

void enqueue(Queue* queue, string data) {
    QNode* newNode = createNode(data);
    if (queue->head == nullptr) {
        queue->head = newNode;
        queue->tail = newNode;
        return;
    }
    
    queue->tail->next = newNode;
    queue->tail = newNode;
}

void dequeue(Queue* queue) {
    if (queue->head == nullptr) {
        cout << "Очередь пустая. Нечего удалять.\n";
        return;
    }

    QNode* deleteNode = queue->head;
    if (deleteNode == queue->tail) {
        queue->head = nullptr;
        queue->tail = nullptr;
    }
    else {
        queue->head = queue->head->next;
    }
    delete deleteNode;
}

void printQueue(Queue* queue) {
    if (queue->head == nullptr) {
        cout << "Очередь пуста.\n";
        return;
    }

    cout << "Очередь: ";
    QNode* curElem = queue->head;
    while (curElem != nullptr) {
        cout << curElem->value << " ";
        curElem = curElem->next;
    }

    cout << endl;
}